### PR TITLE
Ability to override Gerrit URL

### PR DESCRIPTION
**Pull request by djarosz**
- improved change and branch URL's when build based on ChangeEvent.url
- now it is possible to overwrite Gerrit base URL using
  configuration property. This helps if Gerrit is behind proxy
  and external Gerrit URL is different when the one in ChangeEvent.url

### CODE REVIEW

File: app/src/dist/config.yaml [hunk 1]
```diff
@@ -16,9 +16,12 @@ Event buffer timeout in seconds: 15
 #
 # mergedChangeEmoji: If this list is present and non-empty, an emoji from it will be selected for notifications
 # for merged change events. Use if you want to have fun celebrating a change getting merged. :)
+# - gerritUrl: base Gerrit URL used to render change links in Slack notifications. If null base Gerrit url
```
Comment (MichaelSims) [inline app/src/dist/config.yaml:19]: Add comma for clarity: "If null, base Gerrit url..."

File: app/src/main/java/sims/michael/gerritslackbot/EventGroupingTransformer.kt [hunk 2]
```diff
@@ -19,7 +19,7 @@ class EventGroupingTransformer(
         /* Reviewers Added - filters are not applied (messages will be addressed via DM if the reviewer is recognized) */
         val reviewerAddedEvents = events.filterIsInstance(ReviewerAddedEvent::class.java)
         val reviewerAddedMessages = reviewerAddedEvents.groupBy { event ->
-            EventGroup<ReviewerAddedEvent>(
+            EventGroup(
```
Comment (MichaelSims) [inline app/src/main/java/sims/michael/gerritslackbot/EventGroupingTransformer.kt:22]: Good change, but I'd like to see this isolated to a single commit

File: app/src/main/java/sims/michael/gerritslackbot/slack/EventGroupToSlackMessageTransformer.kt [hunk 3]
```diff
@@ -53,7 +53,7 @@ class EventGroupToSlackMessageTransformer(
                 channel = slackName,
                 attachments = listOf(Attachment(
                         fallback = "Summary of changes",
-                        text = events.joinToString("\n") { event -> event.toSlackSummary() }
+                        text = events.map { it.toSlackSummary() }.joinToString("\n")
```
Comment (MichaelSims) [inline app/src/main/java/sims/michael/gerritslackbot/slack/EventGroupToSlackMessageTransformer.kt:56]: Why this change? The previous form is simpler (and IDE inspections suggests it)

File: app/src/main/java/sims/michael/gerritslackbot/slack/EventGroupToSlackMessageTransformer.kt [hunk 4]
```diff
@@ -148,7 +148,7 @@ class EventGroupToSlackMessageTransformer(
     }
 
     private fun EventGroup<*>.toProjectBranchPrefix(): String {
-        val baseUrl = HttpUrl.parse(events.first().change.url).let { "${it.scheme()}://${it.host()}" }
+        val baseUrl =  events.first().gerritBaseUrl()
```
Comment (MichaelSims) [inline app/src/main/java/sims/michael/gerritslackbot/slack/EventGroupToSlackMessageTransformer.kt:151]: Why replace `okhttp3.HttpUrl#parse` with a regex approach? If it's due to needing the port number, that can still be obtained from `okhttp3.HttpUrl`. I'd prefer to use `HttpUrl` because its parsing is much more comprehensive than the regex. Thoughts?

File: app/src/main/java/sims/michael/gerritslackbot/slack/EventGroupToSlackMessageTransformer.kt [hunk 5]
```diff
@@ -174,16 +174,26 @@ class EventGroupToSlackMessageTransformer(
     }
 
     private fun Int.voteToString(): String = (if (this > 0) "+$this" else this.toString()) + "'d"
-    private fun Int.verifyVoteToString(): String = if (this > 0) "Verified (+1)" else "Failed to verify (-1)"
+    private fun Int.verifyVoteToString(): String = when {
```
Comment (MichaelSims) [inline app/src/main/java/sims/michael/gerritslackbot/slack/EventGroupToSlackMessageTransformer.kt:177]: Good change, but I'd prefer to see it isolated to its own commit

File: app/src/main/java/sims/michael/gerritslackbot/slack/EventGroupToSlackMessageTransformer.kt [hunk 6]
```diff
@@ -192,11 +202,4 @@ class EventGroupToSlackMessageTransformer(
                 .joinToString(delimiter).escapeHtml()
     }
 
-    private fun <K : Any?, V> Map<K?, V>.filterNotNullKeys(): Map<K, V> {
```
Comment (MichaelSims) [inline app/src/main/java/sims/michael/gerritslackbot/slack/EventGroupToSlackMessageTransformer.kt:195]: Good change, but again, isolated.

I'm actually OK with all of the "clean up" changes to the existing code being a single commit. That includes this, and the change to drop the explicit type params in `sims.michael.gerritslackbot.EventGroupingTransformer`

File: tests/src/test/resources/sims/michael/gerritslackbot/config.yaml [hunk 7]
```diff
@@ -16,9 +16,12 @@ Event buffer timeout in seconds: 15
 #
 # mergedChangeEmoji: If this list is present and non-empty, an emoji from it will be selected for notifications
 # for merged change events. Use if you want to have fun celebrating a change getting merged. :)
+# - gerritUrl: base Gerrit URL used to render change links in Slack notifications. If null base Gerrit url
```
Comment (MichaelSims) [inline tests/src/test/resources/sims/michael/gerritslackbot/config.yaml:19]: Same comment as above applies here.

**Review by MichaelSims — CHANGES_REQUESTED** [review]
Thanks for this one too. In addition to the inline comments I left (with some questions), again I intend to rebase and in an effort to keep a clean history I'd like each commit to stand on its own as much as is feasible. In this case, we have some changes that are out of band (clean ups) and also some things are introduced in one commit and then later changed in another, and those could be squashed.

As before, I am more than happy to make these changes myself and get it merged but wanted to discuss it first.

Thanks again.

**Comment by MichaelSims** [comment]
I just pushed a change to master that includes all of the "clean up" type changes that are in your two pull requests... Hopefully that makes it easier, not harder...

File: app/src/main/java/sims/michael/gerritslackbot/slack/EventGroupToSlackMessageTransformer.kt (hunk 3 above)
Comment (djarosz) [reply app/src/main/java/sims/michael/gerritslackbot/slack/EventGroupToSlackMessageTransformer.kt:56]: Not sure now but a I thought that a have seen this form in another place of this file. So there were there were two places doing same thing but with different syntax/construction so I just picked one made this parts of code use same construct.

File: app/src/main/java/sims/michael/gerritslackbot/slack/EventGroupToSlackMessageTransformer.kt (hunk 4 above)
Comment (djarosz) [reply app/src/main/java/sims/michael/gerritslackbot/slack/EventGroupToSlackMessageTransformer.kt:151]: My installation uses http reverse proxy so actual URL in received events was some thing like 

http://server:8976/node1/gerrit/4578

So just getting port, host and scheme from okhttp and then adding change number was not working for me. That's why this approach. But maybe this can be done better as I do not know okhttp.

Any way this solution was not enough as http://server:8976 was working inside corporate network and Gerrit installation had different address outside of it. Theta why additional property in config file.

**Comment by djarosz** [comment]
Feel free to make any changes You want....sorry as I'm not able to do them my self for next week.

**Comment by MichaelSims** [comment]
Thanks for the explanations. I understand the issue you were running into regarding the gerrit URL. It's unfortunate that there doesn't seem to be a way to reliably determine the base URL from the event stream. I incorporated your changes slightly differently and added some unit tests for them. Hopefully it addresses your requirements. If not, please let me know and we can revisit. Thanks again for the contributions!
